armsrc/mifaresniff.c

```diff
@@ -1,177 +1,184 @@
-		sniffState = SNF_INIT;
-	}
-
-	switch (sniffState) {
-		case SNF_INIT:{
-			if ((len == 1) && (reader) && (bitCnt == 7) ) {  // REQA or WUPA from reader
-				sniffUIDType = SNF_UID_4;
-				memset(sniffUID, 0x00, 8);
-				memset(sniffATQA, 0x00, 2);
-				sniffSAK = 0;
-				sniffState = SNF_WUPREQ;
-			}
-			break;
-		}
-		case SNF_WUPREQ:{
-			if ((!reader) && (len == 2)) { 		// ATQA from tag
-				memcpy(sniffATQA, data, 2);
-				sniffState = SNF_ATQA;
-			}
-			break;
-		}
-		case SNF_ATQA:{
-			if ((reader) && (len == 2) && (data[0] == 0x93) && (data[1] == 0x20)) { // Select ALL from reader
-				sniffState = SNF_ANTICOL1;
-			}
-			break;
-		}
-		case SNF_ANTICOL1:{
-			if ((!reader) && (len == 5) && ((data[0] ^ data[1] ^ data[2] ^ data[3]) == data[4])) {  // UID from tag (CL1) 
-				memcpy(sniffUID + 3, data, 4);
-				sniffState = SNF_UID1;
-			}
-			break;
-		}
+//-----------------------------------------------------------------------------
+// Merlok - 2012
+//
+// This code is licensed to you under the terms of the GNU GPL, version 2 or,
+// at your option, any later version. See the LICENSE.txt file for the text of
+// the license.
+//-----------------------------------------------------------------------------
+// Routines to support mifare classic sniffer.
+//-----------------------------------------------------------------------------
+
+#include "mifaresniff.h"
+#include "apps.h"
+#include "proxmark3.h"
+#include "util.h"
+#include "string.h"
+#include "iso14443crc.h"
+#include "iso14443a.h"
+#include "crapto1/crapto1.h"
+#include "mifareutil.h"
+#include "common.h"
+
+
+static int sniffState = SNF_INIT;
+static uint8_t sniffUIDType;
+static uint8_t sniffUID[8] = {0x00};
+static uint8_t sniffATQA[2] = {0x00};
+static uint8_t sniffSAK;
+static uint8_t sniffBuf[16] = {0x00};
+static uint32_t timerData = 0;
+
+
+bool MfSniffInit(void){
+	memset(sniffUID, 0x00, 8);
+	memset(sniffATQA, 0x00, 2);
+	sniffSAK = 0;
+	sniffUIDType = SNF_UID_4;
+
+	return FALSE;
+}
+
+bool MfSniffEnd(void){
+	LED_B_ON();
+	cmd_send(CMD_ACK,0,0,0,0,0);
+	LED_B_OFF();
+
+	return FALSE;
+}
+
+bool RAMFUNC MfSniffLogic(const uint8_t *data, uint16_t len, uint8_t *parity, uint16_t bitCnt, bool reader) {
+
+	if (reader && (len == 1) && (bitCnt == 7)) { 		// reset on 7-Bit commands from reader
+		sniffState = SNF_INIT;
+	}
+
+	switch (sniffState) {
+		case SNF_INIT:{
+			if ((len == 1) && (reader) && (bitCnt == 7) ) {  // REQA or WUPA from reader
+				sniffUIDType = SNF_UID_4;
+				memset(sniffUID, 0x00, 8);
+				memset(sniffATQA, 0x00, 2);
+				sniffSAK = 0;
+				sniffState = SNF_WUPREQ;
+			}
+			break;
+		}
+		case SNF_WUPREQ:{
+			if ((!reader) && (len == 2)) { 		// ATQA from tag
+				memcpy(sniffATQA, data, 2);
+				sniffState = SNF_ATQA;
+			}
+			break;
+		}
+		case SNF_ATQA:
 		case SNF_UID1:{
+			// SNF_ATQA
+			if ((reader) && (len == 2) && (data[0] == 0x93) && (data[1] == 0x20)) { // Select ALL from reader
```

Comment:
Idea: can't we simply ignore the SELECT ALL command and wait for the SELECT UID instead? For both CL1 and CL2 ?

Reply:
we cant because the card answers to this command, but we can do nothing)

Reply:
We are passively sniffing. The card will answer to SELECT ALL and the reader will then issue a SELECT UID.  The `MfSniffLogic()` will be called for each reader command and tag response. But there is no need to have a respective `case` or `if` for each of them. Some commands or tag responses may simply "fall through". We would only need to handle `SNF_UID1` and `SNF_UID2` with the 0x93 0x70 and 0x95 0x70 commands in order to get the UID.

We could even skip the ATQA and SAK cases if we wouldn't want to display them on client side (would need another indicator then for 4 or 7 byte UID though).

Every tag response we are *not* checking makes `hf mf sniff` more robust.

Or do I miss something?

Reply:
yes, we can just skip this command.
ill try

Reply:
works